design/envoy-json-logging-custom-fields-design.md

```diff
@@ -0,0 +1,144 @@
+
+Status: Draft
+
+## Abstract
+Allow users to add arbitrary supported Envoy JSON fields to the access log.
+
+## Background
+This feature was requested in [#3032](https://github.com/projectcontour/contour/issues/3032).
+Contour [allows Envoy to output JSON logs](https://github.com/projectcontour/contour/blob/main/design/envoy-json-logging.md).
+The current opinionated implementation allows users to configure log fields from the [limited list](https://godoc.org/github.com/projectcontour/contour/internal/envoy#DefaultFields) of available Envoy fields.
+
+The reason that the current implementation restricts users to an approved list of fields is that recovering from sending invalid config to Envoy is hard.
+If invalid JSON fields are included in the Envoy config, then Envoy will reject the listener and will not accept updates from Contour until the Envoy is restarted (just removing the error will not fix the issue).
+
+## Goals
+* Support nearly all the fields listed in [the Envoy docs][1]
+* Specifically, support configuration of parameterized JSON fields
+* Validation of the JSON fields config
+* Continue to support existing `json-fields` translations (existing config continues to work)
+
+## Non Goals
+* Supporting `DYNAMIC_METADATA` and `FILTER_STATE`.
+  These parameterized fields are more complicated to validate.
+
+
+## High-Level Design
+
+* The current `json-fields` Contour configuration will be extended to accept values of the form: `field_name="literal string with Envoy field operator"`.
+
+
+## Detailed Design
+
+### Configuration
+
+The current `json-fields` configuration accepts the list of fields from the predefined list of available translations, defined in the [source code](https://github.com/projectcontour/contour/blob/main/internal/envoy/accesslog.go#L19).
+The configuration will be extended to also accept values in the form of: `field_name="literal string with Envoy field operator"`.
+
+The value will be parsed in the following manner:
+
+* Everything before the first `=` character will be treated as the JSON field name
+* Everything after the first `=` character will be treated as Envoy field [format command][1]
+
+If there is no `=`, then the whole value is the field name. The field will be translated using the already defined translation map.
+
+### Validation
+
+The Envoy field command will be validated to prevent Envoy misconfiguration.
+The validation will be performed using regular expressions.
+These expressions will be similar to [one used in Envoy](https://github.com/envoyproxy/envoy/blob/4d77fc802c3bc1c517e66c54e9c9507ed7ae8d9b/source/common/formatter/substitution_formatter.cc#L291), but with the specification of exact commands.
+If the validation is failed an error will be raised and printed in the logs of Contour during startup.
+
+### Edge Cases
+
+#### Field name is repeated
+
+If the field name is repeated the last entry in the configuration takes precedes and overwrites the previous value.
```

Comment:
Wonder if this should raise a validation error instead of overwriting?

Reply:
A validation error could be OK if a field name is repeated in a single `json-fields` array. My original intention with suggesting overwriting was that operators should be able to use this mechanism to change the definition of builtin log fields (ie. overwriting the builtin field mapping).